File: src/EcommerceDDD.Spa/src/app/modules/ecommerce/models/Product.ts
export class Product {
  productId: string;
  name: string;
  description: string;
  category: string;
  imageUrl: string;
  unitPrice: number;
  currency: string;
  currencySymbol: string;
  quantity: number;
  quantityInStock: number;

  constructor(
    productId: string,
    name: string,
    description: string,
    category: string,
    imageUrl: string,
    unitPrice: number,
    currency: string,
    currencySymbol: string,
    quantity: number,
    quantityInStock: number
  ) {
    this.productId = productId;
    this.name = name;
    this.description = description;
    this.category = category;
    this.imageUrl = imageUrl;
    this.unitPrice = unitPrice;
    this.currency = currency;
    this.currencySymbol = currencySymbol;
    this.quantity = quantity;
    this.quantityInStock = quantityInStock;
  }
}

File: src/Services/EcommerceDDD.ProductCatalog/Domain/Product.cs
namespace EcommerceDDD.ProductCatalog.Domain;

public class Product : AggregateRoot<ProductId>
{
    public string Name { get; private set; }
    public string Category { get; private set; }
    public string Description { get; private set; }
    public string ImageUrl { get; private set; }
    public Money UnitPrice { get; private set; }

    public static Product Create(ProductData productData, Guid? productId = null)
    {
        var (Name, Category, Description, ImageUrl, UnitPrice) = productData
            ?? throw new ArgumentNullException(nameof(productData));

        if (string.IsNullOrWhiteSpace(Name))
            throw new DomainException("Product name cannot be null or whitespace.");
        if (string.IsNullOrWhiteSpace(Category))
            throw new DomainException("Product category cannot be null or whitespace.");
        if (string.IsNullOrWhiteSpace(Description))
            throw new DomainException("Product description cannot be null or whitespace.");
        if (string.IsNullOrWhiteSpace(ImageUrl))
            throw new DomainException("ImageUrl cannot be null.");
        if (UnitPrice is null)
            throw new DomainException("Product unit price cannot be null.");

        return new Product(productData, productId);
    }

    private Product(ProductData productData, Guid? productId = null)
    {
        Id = ProductId.Of(productId ?? Guid.NewGuid());
        Name = productData.Name;
        Category = productData.Category;
        Description = productData.Description;
        ImageUrl = productData.ImageUrl;
        UnitPrice = productData.UnitPrice;
    }

    private Product(){}
}